### ISSUE TITLE
Regarding Training / Fine-tuning YOLO-E and YOLO-W models

### ISSUE THREAD

**Issue by ArghyaChatterjee** [ultralytics/ultralytics#25553]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am trying to increase the annotation for training by doing one class at a time. Meaning I am only focusing on one class in a single image and ignoring other classes which I also want the detector to detect. Now this is not valid for closed vocabulary models like YOLOv8 / YOLOv11 or similar models since the detector will learn that one labeled class and unlearn the other class/classes which are not labeled. However, is the above process same for open vocabulary models like YOLO-E or YOLO-W or they can learn from one class at a time without forgetting other classes ?

This is a scalability issue since if expanding the annotated dataset mean labeling all the classes in all the previous images, it's almost impossible to go back and label new classes which was not labeled initially.

Thanks in advance.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ArghyaChatterjee, thank you for your interest in Ultralytics 🚀! This is an automated response, and an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many common questions may already be answered.

For your question about training and fine-tuning YOLO-E or YOLO-W, please provide as much information as possible, including the exact model and package version, dataset structure and labeling format, training command or code, representative dataset images, and training logs. Please also verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips). If you can reproduce the behavior with a small example, a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) would be especially helpful 🧪

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Yes—the same caveat applies during standard YOLOE/YOLO-World fine-tuning: open-vocabulary prompting does not make training partial-label-aware, so desired objects left unlabeled may be learned as background and existing capabilities may degrade. Use fully labeled training images, exclude incompletely labeled images, or generate and review pseudo-labels; also mix representative old-class data during incremental fine-tuning. If zero-shot accuracy is sufficient, use [YOLOE prompts](https://docs.ultralytics.com/models/yoloe) or [YOLO-World custom prompts](https://docs.ultralytics.com/models/yolo-world#set-prompts) without retraining.

**Comment by Y-T-G** [comment]
The model trains all the weights during training. There are no class specific weights. So it's not possible to freeze the performance on certain classes while training others.

**Comment by ArghyaChatterjee** [comment]
Thanks for the reply. I had the impression that YOLO-E and YOLO-W already can recognize a lot of the things/classes and I can gradually train or finetune it with new annotated images / classes which it thinks to be something as part of open vocabulary training but after finetuning it will say that as that specifically. Isn't this the case ? I thought I am training / finetuning specific things and the open vocabulary model can still detect other classes inside the image that it's already trained but it's just that specific class is well detected and detected with the name (say instead of saying black box, it will say it as storage container or black storage container as I trained it with that class label). Isn't it the case ? If that's the case, then what's the problem of labeling one class at a time. I am just finetuning a set of objects with specific names and classes in the whole set. During finetuning, I won't expect that to forget the whole set and focus only on my set.

**Comment by Y-T-G** [comment]
Catastrophic forgetting is an unsolved problem.

Gradient descent only optimizes for what's in the batch. It has no separate memory for "don't forget old classes" unless old classes are in the data.

All weights are shared across all classes. There's no dedicated block of weights per class (except the last layer; which is addresed below). So any update that helps the new class can also shift weights that the old classes depend on, even with backbone frozen. Freezing the backbone just reduces how many weights are shared and touched. The head that's still training is still shared across every class.

The best you could do is linear probing which the exploits the last layer caveat I mentioned:

For this to work, you need two things:

1. Always have the list of classes you want to detect in your dataset YAML. You can't only have the classes that you want to detect. You need all the classes in the YAML even if you're not training on them currently.

2. The model should already have the required features in the backbone. Because everything other than the last layer will be frozen. So it can't learn to extract new features. It's a very limited fine-tuning technique. YOLOE fine-tuning would initialize with the CLIP weights using the class names in your YAML file as prompts before it starts, so it will have some idea about the class.

Even the last layer isn't fully safe by default though, backbone-freezing stops other classes' features from moving, but the last conv layer still has one row of weights per class, and a plain linear-probe training run lets every row in it update, not just the new class's. So on top of freezing everything before the head, you also want to lock every row except the new class's:

```python
import torch
from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOEPETrainer

NEW_CLASS_INDEX = 0  # position of your new class in data.yaml's `names`

def _mask_grad(grad, idx):
    m = torch.zeros_like(grad)  # built from `grad`, so always the right device
    m[idx] = 1.0
    return grad * m


class YOLOESingleClassTrainer(YOLOEPETrainer):
    def get_model(self, cfg=None, weights=None, verbose=True):
        model = super().get_model(cfg, weights, verbose)
        head = model.model[-1]

        cv3_branches = [head.cv3]
        if getattr(head, "one2one_cv3", None) is not None:
            cv3_branches.append(head.one2one_cv3)

        self._locked_convs = []
        for cv3 in cv3_branches:
            for seq in cv3:  # one per detection scale (P3, P4, P5)
                conv = seq[-1]  # fused Conv2d(c3, num_classes, 1)

                # snapshot as buffers, not raw tensors — buffers move with
                # the module when the trainer calls model.to(device);
                # persistent=False keeps them out of saved checkpoints
                conv.register_buffer("orig_weight", conv.weight.detach().clone(), persistent=False)
                conv.register_buffer("orig_bias", conv.bias.detach().clone(), persistent=False)

                conv.weight.register_hook(lambda g: _mask_grad(g, NEW_CLASS_INDEX))
                conv.bias.register_hook(lambda g: _mask_grad(g, NEW_CLASS_INDEX))

                self._locked_convs.append(conv)
        return model

    def optimizer_step(self):
        self.scaler.unscale_(self.optimizer)
        torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)
        self.scaler.step(self.optimizer)
        self.scaler.update()
        self.optimizer.zero_grad()

        with torch.no_grad():
            for conv in self._locked_convs:
                # re-fetch conv.weight / conv.orig_weight fresh each step —
                # both live on the module, so both reflect the current device
                mw = torch.zeros_like(conv.weight)
                mw[NEW_CLASS_INDEX] = 1.0
                mb = torch.zeros_like(conv.bias)
                mb[NEW_CLASS_INDEX] = 1.0
                conv.weight.mul_(mw).add_(conv.orig_weight * (1 - mw))
                conv.bias.mul_(mb).add_(conv.orig_bias * (1 - mb))

        if self.ema:
            self.ema.update(self.model)

model = YOLOE("yoloe-26s.yaml")
model.load("yoloe-26s-seg.pt")

head_index = len(model.model.model) - 1
freeze = [str(i) for i in range(head_index)]  # backbone + neck
for name, _ in model.model.model[-1].named_children():
    if "cv3" not in name:
        freeze.append(f"{head_index}.{name}")
    else:
        for s in range(3):  # P3, P4, P5
            freeze.append(f"{head_index}.{name}.{s}.0")
            freeze.append(f"{head_index}.{name}.{s}.1")

results = model.train(
    data="path/to/data.yaml",
    epochs=80,
    patience=10,
    trainer=YOLOESingleClassTrainer,
    freeze=freeze,
)
```

You wouldn't be able to escape overfitting. The model is still learning on your dataset and your dataset only. It only cares about the performance on your dataset, and that means it can lose performance on the other dataset.

**Comment by ArghyaChatterjee** [comment]
Thanks for the explanation. This helps a lot.
